### PR TITLE
#[app] proc-macro

### PR DESCRIPTION
**Pull request by YaronWittenstein**
# Motivation

This PR pretty much reimplements the `svm-sdk-macros` crate.
Firstly, it introduces a new proc-marco named, `#[app]` that serves as the root macro.
It decorates a module that implements SVM app, and they're also sub-proc-macros underneath.

Part of these sub-macros already exists (such as `#[storage]` and `[endpoint]`) and new ones are added. 

This big rewrite should make the amount of work required for adding new proc-macros in the future much easier and faster.

### CODE REVIEW

File: crates/svm-sdk-tests/tests/endpoint/endpoint_with_fundable_hook.stderr [hunk 1]
```diff
@@ -0,0 +1,7 @@
+error: #[endpoint]` and `#[fundable_hook]` can't co-exist.
```
[the remark this answers is not in the document]
Comment (YaronWittenstein) [reply crates/svm-sdk-tests/tests/endpoint/endpoint_with_fundable_hook.stderr:1]: it's auto-generated by the `trybuild`.
The first time you run a test it's generated with a warning
and it asks you to move the error to the same directory as the test.

File: crates/svm-abi/decoder/src/decoder.rs [hunk 2]
```diff
@@ -19,6 +19,7 @@ pub enum TypeError {
 
 enum TypeKind {
     None,
+    Unit,
```
[the remark this answers is not in the document]
Comment (YaronWittenstein) [reply crates/svm-abi/decoder/src/decoder.rs:22]: `()`

[the remark this answers is not in the document]
Comment (YaronWittenstein) [reply crates/svm-abi/decoder/src/decoder.rs:22]: A function in Rust that returns nothing actually returns `()`

**Review by moshababo — APPROVED** [review]

File: crates/svm-sdk-tests/tests/endpoint/funding.rs [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    #[storage]
+    struct Storage {
+        coins: Amount,
+
+        called: bool,
+    }
+
+    #[fundable(update_coins)]
+    #[endpoint]
+    fn do_nothing() {}
```
Comment (YaronWittenstein) [inline crates/svm-sdk-tests/tests/endpoint/funding.rs:17]: This `endpoint` returns `()` 
@moshababo

Comment (moshababo) [reply crates/svm-sdk-tests/tests/endpoint/funding.rs:17]: Got it, 10x

**Pull request merged**